### PR TITLE
ARROW-605: [C++] Refactor IPC adapter code into generic ArrayLoader class. Add Date32Type

### PR DESCRIPTION
**Pull request by wesm**
These are various changes introduced to support the Feather merge in ARROW-452 #361

### CODE REVIEW

File: cpp/src/arrow/type.h [hunk 1]
```diff
@@ -467,16 +486,20 @@ struct ARROW_EXPORT TimestampType : public FixedWidthType {
 
   int bit_width() const override { return static_cast<int>(sizeof(int64_t) * 8); }
 
-  TimeUnit unit;
-
   explicit TimestampType(TimeUnit unit = TimeUnit::MILLI)
       : FixedWidthType(Type::TIMESTAMP), unit(unit) {}
 
+  explicit TimestampType(const std::string& timezone, TimeUnit unit = TimeUnit::MILLI)
```
Comment (tebeka) [inline cpp/src/arrow/type.h]: Shouldn't `timezone` default to `"UTC"`? Also maybe it should be the 2nd parameter?

Comment (wesm) [inline cpp/src/arrow/type.h]: We haven't resolved any questions about storing time zones in the IPC metadata (https://github.com/apache/arrow/blob/master/format/Message.fbs), so I did the bare minimum here to support the pandas and R use cases, where they have the notion of tz-naive and tz-aware timestamps. So I don't want to invest any more energy in it until we have a broader discussion about time zones

**Comment by wesm** [comment]
fixing compiler warnings

**Comment by wesm** [comment]
green build: https://travis-ci.org/wesm/arrow/builds/209347029

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -67,9 +67,12 @@ struct Type {
     // Variable-length bytes (no guarantee of UTF8-ness)
     BINARY,
 
-    // By default, int32 days since the UNIX epoch
+    // int64_t milliseconds since the UNIX epoch
     DATE,
 
+    // int32_t days since the UNIX epoch
+    DATE32,
```
Comment (xhochy) [inline cpp/src/arrow/type.h]: Is the ordering of the types important yet, i.e. do we match the enums anywhere?

File: cpp/src/arrow/type.h (hunk 1 above)
Comment (xhochy) [inline cpp/src/arrow/type.h]: Empty string thus indicates a tv-naive timestamp?

**Review by xhochy — APPROVED** [review]
Except the two questions, this looks good. If they both can be answered with yes, feel free to merge.

File: cpp/src/arrow/type.h (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/type.h]: Right, wasn't sure what else to use for a "null" value at the moment

File: cpp/src/arrow/type.h (hunk 2 above)
Comment (wesm) [inline cpp/src/arrow/type.h]: I don't think so, and I don't think we should guarantee ABI stability of the enum values yet anyhow (at some point we will want to)

**Comment by wesm** [comment]
thanks

**Pull request closed without merging**